File: server/communication/message.py
LENGTH_HEADER_SIZE = 4

class ConfirmationMessage:
    def __init__(self, status):
        """Creates a new ConfirmationMessage instance"""
        self.status = status

def send_message(socket, data):
    """Sends a message through the socket"""
    total_bytes_sent = 0
    while total_bytes_sent < len(data):
        bytes_sent = socket.send(data[total_bytes_sent:])
        if bytes_sent == 0:
            raise ConnectionError
        total_bytes_sent += bytes_sent
    return total_bytes_sent

def read_message(socket, length):
    """Reads a message from the socket"""
    data = bytearray()
    total_bytes_received = 0
    while total_bytes_received < length:
        bytesReceived = socket.recv(length - total_bytes_received)
        if not bytesReceived:
            raise ConnectionError
        data.extend(bytesReceived)
        total_bytes_received += len(bytesReceived)
    return data

def receive_message(socket):
    """Receives the payload of a message and its type from the socket"""
    length_bytes = read_message(socket, LENGTH_HEADER_SIZE)
    message_length = int.from_bytes(length_bytes, byteorder='big')
    full_message = read_message(socket, message_length)
    message_type = full_message[0]
    payload = full_message[1:]
    return message_type, payload

File: client/communication/message.go
package communication

import (
    "net"
    "bytes"
    "encoding/binary"
    "fmt"
)

const LengthHeaderSize = 4
const MaxNameLength = 50

// BetMessage represents the message sent by the client to register a bet
type BetMessage struct {
    Agency    string
    FirstName string
    LastName  string
    Document  string
    Birthdate string
    Number    string
}

// ConfirmationMessage represents the confirmation message received by the client
type ConfirmationMessage struct {
    Status string
}

// NewBetMessage creates a new BetMessage instance
func NewBetMessage(agency, firstName, lastName, document, birthdate, number string) (BetMessage, error) {
    if len(firstName) > MaxNameLength || len(lastName) > MaxNameLength {
        return BetMessage{}, fmt.Errorf("nombre y apellido no pueden tener más de %d caracteres", MaxNameLength)
    }
    return BetMessage{
        Agency:    agency,
        FirstName: firstName,
        LastName:  lastName,
        Document:  document,
        Birthdate: birthdate,
        Number:    number,
    }, nil
}

// NewConfirmationMessage creates a new ConfirmationMessage instance
func NewConfirmationMessage(status string) ConfirmationMessage {
    return ConfirmationMessage{
        Status: status,
    }
}

// SendMessage sends a message through the connection
func SendMessage(conn net.Conn, data []byte) error {
    totalBytesSent := 0
    for totalBytesSent < len(data) {
        bytesSent, err := conn.Write(data[totalBytesSent:])
        if err != nil {
            return err
        }
        totalBytesSent += bytesSent
    }
    return nil
}

// ReceiveMessage receives a message through the connection
func ReceiveMessage(conn net.Conn, length uint32) ([]byte, error) {
    totalBytesReceived := 0
    data := make([]byte, length)
    for totalBytesReceived < int(length) {
        bytesReceived, err := conn.Read(data[totalBytesReceived:])
        if err != nil {
            return nil, err
        }
        totalBytesReceived += bytesReceived
    }
    return data, nil
}

// ReadMessagePayload reads the payload of a message from the connection
func ReadMessagePayload(conn net.Conn) ([]byte, error) {
    lengthBytes, err := ReceiveMessage(conn, LengthHeaderSize)
    if err != nil {
        return nil, err
    }

    var length uint32
    buf := bytes.NewReader(lengthBytes)
    if err := binary.Read(buf, binary.BigEndian, &length); err != nil {
        return nil, err
    }

    payloadBytes, err := ReceiveMessage(conn, length)
    if err != nil {
        return nil, err
    }
    
    return payloadBytes, nil
}
